client-tools/Shell/TelemetricsHandler.cpp

```diff
@@ -0,0 +1,348 @@
+// is sent to an endpoint, though not exactly the same thing, because here the
+// endpoint is local, hence the same client is used and a new endpoint is not
+// created.
+std::optional<VPackBuilder> TelemetricsHandler::sendTelemetricsToEndpoint(
+    std::string const& reqUrl) {
+  VPackBuilder testResponseBuilder;
+
+  bool isLocal = false;
+  if (reqUrl.starts_with('/')) {
+    isLocal = true;
+  }
+  bool isOriginalUrl = (reqUrl.compare(kOriginalUrl) == 0) ? true : false;
+
+  std::string url = reqUrl;
+
+  Result result = {TRI_ERROR_NO_ERROR};
+  uint32_t timeoutInSecs = 10;
+  uint64_t sslProtocol = TLS_V13;
+  uint32_t maxRedirects = 5;
+
+  std::string const body = _telemetricsFetchedInfo.toJson();
+
+  uint32_t numRedirects = 0;
+  basics::StringBuffer compressedBody;
+
+  while (!_server.isStopping()) {
+    std::unique_lock lk(_mtx, std::defer_lock);
+    std::unordered_map<std::string, std::string> headers;
+    headers.emplace(StaticStrings::ContentTypeHeader,
+                    StaticStrings::MimeTypeJson);
+    std::string relativeUrl;
+    if (!isOriginalUrl) {
+      headers.emplace(StaticStrings::ContentLength,
+                      std::to_string(compressedBody.size()));
+    } else {
+      auto res =
+          encoding::gzipCompress(reinterpret_cast<uint8_t const*>(body.data()),
+                                 body.size(), compressedBody);
+      if (res != TRI_ERROR_NO_ERROR) {
+        break;
+      }
+      headers.emplace(StaticStrings::ContentLength,
+                      std::to_string(body.size()));
+      headers.emplace(StaticStrings::ContentEncoding,
+                      StaticStrings::EncodingGzip);
+    }
+    if (isLocal) {
+      relativeUrl = url;
+      ClientManager clientManager(
+          _server.getFeature<HttpEndpointProvider, ClientFeature>(),
+          Logger::FIXME);
+      lk.lock();
+      _httpClient = clientManager.getConnectedClient(true, false, false, 0);
+    } else {
+      std::string lastEndpoint = getEndpointFromUrl(url);
+      std::vector<std::string> endpoints;
+      auto [endpoint, relative, error] =
+          getEndpoint(nullptr, endpoints, url, lastEndpoint);
+      relativeUrl = std::move(relative);
+      ClientFeature& client =
+          _server.getFeature<HttpEndpointProvider, ClientFeature>();
+      std::unique_ptr<Endpoint> newEndpoint(Endpoint::clientFactory(endpoint));
+
+      lk.lock();
+
+      TRI_ASSERT(newEndpoint.get() != nullptr);
+      std::unique_ptr<httpclient::GeneralClientConnection> connection(
+          httpclient::GeneralClientConnection::factory(
+              client.getCommFeaturePhase(), newEndpoint, 30, 60, 3,
+              sslProtocol));
+
+      if (connection == nullptr) {
+        continue;
+      }
+      httpclient::SimpleHttpClientParams const params(30, false);
+      _httpClient = std::make_unique<httpclient::SimpleHttpClient>(
+          connection.get(), params);
```

Comment:
Passing `connection.get()` to the SimpleHttpClient is problematic here, because the SimpleHttpClient then will not assume ownership of the connection. It will just store and use the raw pointer to the connection.
However, after line 230 the connection object will go out of scope and will be deleted. 
So the SimpleHttpClient will end up with a pointer to a deleted connection. UB.